internal/pkg/levee/levee.go

```diff
@@ -79,16 +102,114 @@ func run(pass *analysis.Pass) (interface{}, error) {
+	}
+
+	for _, v := range vs {
+		// If we ask the pointer package to analyze a value it can't analyze (i.e., not a pointer or
+		// pointer-like type), the analysis will fail and produce error output in stderr.
+		if !pointer.CanPoint(v.Type()) {
+			continue
+		}
+		// Let the configuration know that we are interested in computing this value's points-to set.
+		// The points-to set is the set of values that this pointer value might point to.
+		pointerConf.AddQuery(v)
+	}
+
+	// Perform the actual analysis, in accordance with our configuration.
+	// The pointer package's documentation says that "an error can occur only due to an internal bug".
+	// However, failing to provide a main package, or querying for a non-queryable value, will both lead
+	// to errors. We prevent these cases with explicit checks above.
+	result, err := pointer.Analyze(pointerConf)
+	if err != nil {
+		return false
+	}
+
+	// A call was analyzed successfully if performing pointer analysis on it led to a report.
+	analyzedSuccesfully = false
+	for _, ptr := range result.Queries {
+		// Obtain the points-to set for this pointer. Labels are used to get at the actual ssa.Values
+		// that are pointed to.
+		labels := ptr.PointsTo().Labels()
+		if len(labels) == 0 {
+			continue
+		}
+		for _, lab := range labels {
+			v := lab.Value()
+			switch t := v.(type) {
+			case *ssa.MakeInterface:
+				v = t.X
+			}
+			if !analysisConf.IsSource(utils.Dereference(v.Type())) {
+				continue
+			}
+			analyzedSuccesfully = true
+			reportAtSource(pass, v.(ssa.Node), (ssa.Node)(c))
+		}
+	}
+	return analyzedSuccesfully
+}
+
 func makeCall(c *ssa.Call) call.Call {
 	if sinkVarargs := varargs.New(c); sinkVarargs != nil {
 		return sinkVarargs
 	}
 	return call.Regular(c)
 }
 
-func report(pass *analysis.Pass, source *source.Source, sink ssa.Node) {
+func reportAtSink(pass *analysis.Pass, source ssa.Node, sink ssa.Node) {
 	var b strings.Builder
-	b.WriteString("a source has reached a sink")
-	fmt.Fprintf(&b, ", source: %v", pass.Fset.Position(source.Node().Pos()))
+	b.WriteString(sourceReachedSinkMessage)
+	fmt.Fprintf(&b, ", source: %v", pass.Fset.Position(source.Pos()))
 	pass.Reportf(sink.Pos(), b.String())
 }
+
+func reportAtSource(pass *analysis.Pass, source ssa.Node, sink ssa.Node) {
+	var b strings.Builder
+	b.WriteString(sourceReachedSinkMessage)
+	fmt.Fprintf(&b, ", sink: %v", pass.Fset.Position(sink.Pos()))
+	pass.Reportf(source.Pos(), b.String())
+}
+
+// callValues collects the ssa.Values that are arguments to an ssa.Call.
+// In particular, it handles variadic arguments to some extent.
+// We don't want to analyze what the variadic argument points to (a slice), but we do
+// want to analyze the values *within* the slice.
+func callValues(c *ssa.Call) (values []ssa.Value) {
+	for _, a := range c.Call.Args {
+		slice, ok := a.(*ssa.Slice)
+		if !ok {
+			values = append(values, a)
+			continue
+		}
+		refs := *slice.X.Referrers()
+		// the last referrer is the slice
```

Comment:
This feels imprecise.  We have a `slice` above and are examining its referrers.  The last referrer of the slice is the slice?  And that's why we're only iterating to `len(refs)-1`?

I thought `Referrers()` order wasn't guaranteed.  A simple equality check and `continue` within the loop would be more explicit, and would resolve immutableT's concern about emptyness.

Reply:
> The last referrer of the slice is the slice? And that's why we're only iterating to len(refs)-1?

Correct.

> I thought Referrers() order wasn't guaranteed.

I couldn't find anything in the `ssa` package's docs that explicitly says that the order of `Referrers()` is not guaranteed. 

> A simple equality check and continue within the loop would be more explicit, and would resolve immutableT's concern about emptyness.

Agreed. Done.